packages/flow-log/src/components/f-log/f-log-search-and-filters.ts

```diff
@@ -0,0 +1,126 @@
+import { FLog } from "./f-log";
+import * as Mark from "mark.js";
+
+export function closeSearchBar(this: FLog) {
+	this.showToolbar = false;
+	this.highlightText("");
+}
+
+/**
+ * Filter lines based on selected log level
+ * @param filterText
+ */
+export function filterLines(this: FLog, filterText: string) {
+	if (filterText.trim() !== "") {
+		const lines = this.shadowRoot?.querySelectorAll(".log-line");
+		if (lines) {
+			lines.forEach(element => {
+				if (!element.textContent?.toLowerCase()?.includes(filterText)) {
+					element.classList.add("hidden");
+				}
+			});
+		}
+	}
+}
+/**
+ * clear filtered lines
+ */
+export function clearFilter(this: FLog) {
+	const lines = this.shadowRoot?.querySelectorAll(".log-line.hidden");
+	if (lines) {
+		lines.forEach(element => {
+			element.classList.remove("hidden");
+		});
+	}
+}
+
+/**
+ * It will highlight text present in logs with mark tag
+ * @param searchText
+ */
+export function highlightText(this: FLog, searchText: string): void {
+	if (this.searchDebounceTimeout) {
+		clearTimeout(this.searchDebounceTimeout);
+	}
+	this.searchDebounceTimeout = setTimeout(() => {
+		const markInstance = new Mark(
+			this.shadowRoot?.querySelectorAll(".log-line:not(.hidden)") as NodeListOf<HTMLSpanElement>
+		);
+		if (this.lastSearchValue && this.lastSearchValue !== "") {
+			markInstance.unmark(this.lastSearchValue as unknown as HTMLElement);
+		}
+		this.lastSearchValue = searchText;
+		if (searchText && searchText !== "") {
+			markInstance.mark(searchText, {
+				done: (occurrences: number) => {
+					this.searchOccurrences = occurrences;
+					const firstMark = this.shadowRoot?.querySelector("mark[data-markjs='true']");
+					firstMark?.scrollIntoView({ block: "start", behavior: "smooth" });
+					firstMark?.classList.add("active");
+					this.currentMarkIndex = 0;
+					if (occurrences > 0 && this.searchInput) {
+						this.searchInput.suggestElement.suffix = `1 of ${occurrences}`;
+					} else if (this.searchInput) {
+						this.searchInput.suggestElement.suffix = `No results`;
+					}
+				}
+			});
+		} else {
+			if (this.searchInput) this.searchInput.suggestElement.suffix = undefined;
+			this.searchOccurrences = 0;
+		}
+	}, 500);
+}
+
+/**
+ * when next arrow is clicked, it will go to next highlighted mark
+ */
+export function nextMark(this: FLog) {
+	if (this.searchOccurrences > 0 && this.allMarks) {
```

Comment:
Where is `allMarks` defined? I can't see it

Reply:
It is in `f-log.ts` and defined using `@queryAll` decorator
<img width="471" alt="Screenshot 2023-12-13 at 10 51 44 AM" src="https://github.com/cldcvr/flow-core/assets/67629551/4c9a9267-b07d-4ac0-acac-ee5cba8ee1e0">